docs: update logs documentation (#9167)
supabase/supabase: apps/reference/docs/guides/platform/logs.mdx

description: Getting started with Supabase Platform Log Browser
---

> The Supabase Platform provides a log explorer that allows log tracing and debugging. The availability of logs are based on your plan's log retention, which can be found on the [pricing page](https://supabase.com/pricing).

Reviewer: ```suggestion
The Supabase Platform provides a log explorer that allows log tracing and debugging. The availability of logs are based on your Project's billing [plan](https://supabase.com/pricing).
```

:::note

> The features discussed in this article are only available through the **Supabase Platform** and are not yet available on self-hosted.

Reviewer: ```suggestion
The features discussed in this article are available through the **Supabase Platform** and are not yet available on self-hosted.
```

:::

import Tabs from '@theme/Tabs';

> import TabItem from '@theme/TabItem';

Reviewer: Move these to the top

## Product Logs

Supabase provides a logging interface specific to each product.

Simple regex searching is available through the search bar to search the log event messsage for specific keywords and patterns.

> You may also export and download the logs as a spreadsheet.

Author: it would actually export out whatever is fetched, so if user clicks on "Load older", it'll be >100. I'll leave the 100 value out of the sentence.

<!-- 
To update the screenshots, ensure that at least one log line is selected to display the metadata. Can use meme.town as an exmaple.
-->

<Tabs>
  <TabItem value="api" label="API" default>

The API Logs can be found under `Database > API Logs`. These show all network requests for the REST and GraphQL [API](../../guides/api).

![API Logs](/img/guides/platform/logs/logs-api.png)

  </TabItem>
  <TabItem value="postgres" label="Postgres">

The Postgres Logs can be found under `Database > Postgres Logs`. These show all queries and activity for your [Database](../../guides/database).

![Postgres Logs](/img/guides/platform/logs/logs-database.png)

  </TabItem>
  <TabItem value="auth" label="Auth">

The Auth Logs can be found under `Authentication > Logs`. These show all server logs for your [Auth system](../../guides/auth).

![Auth Logs](/img/guides/platform/logs/logs-auth.png)

  </TabItem>
  <TabItem value="storage" label="Storage">

> The Storage Logs can be found under `Storage > Logs`. These show all server logs for your [Storage API](../../guides/storage).

Author: we are enabling it this week. You can add yourself to the supabase team prod segement in configcat, [see here](https://www.notion.so/supabase/Feature-Flags-70e49d813bf74d1597dc68c75af9f967#3cb3d1fe359a4f2e9cead3383072f921)

![Storage Logs](/img/guides/platform/logs/logs-storage.png)

  </TabItem>
  <TabItem value="realtime" label="Realtime">

The Realtime Logs can be found under `Database > Realtime Logs`. These show all server logs for your [Realtime API](../../guides/realtime).

![Realtime Logs](/img/guides/platform/logs/logs-realtime.png)

  </TabItem>
  <TabItem value="functions" label="Edge Functions">

The Edge Functions Logs can be found under two tabs:
- `Functions > Your Function > Invocations`: Edge logs of function calls.
- `Functions > Your Function > Logs`: Logs emitted during function execution.

![Function Edge Logs](/img/guides/platform/logs/logs-functions-edge.png)

![Function Logs](/img/guides/platform/logs/logs-functions.png)

  </TabItem>
</Tabs>

## Logs Explorer

> ![Logs Explorer](/img/guides/platform/logs/logs-explorer.png)

Author: Shifted down

> The log browser can be accessed in the sidebar under **Logs Explorer**. The **Logs Explorer** is for querying and aggregating project logs across products using SQL `SELECT` queries.

Reviewer: ```suggestion
The Dashboard includes a **Logs Explorer** for querying and aggregating logs across products using SQL `SELECT` queries. This can be accessed in the sidebar under **Logs Explorer**. 
```

> The log browser can be accessed in the sidebar under **Logs Explorer**. The **Logs Explorer** is for querying and aggregating project logs across products using SQL `SELECT` queries.

Author: no diff. deleting

> The Logs Explorer exposes logs from each part of the Supabase stack as a separate table that can be queried and joined using SQL.

Reviewer: This is repeating what is said above ...

> The following log sources are currently supported and can be viewed under the **Sources** dropdown:

Reviewer: ```suggestion
The following log sources can be accessed under the **Sources** drop-down:
```

- `edge_logs`: Edge network logs, containing request and response metadata retrieved from Cloudflare.
- `postgres_logs`: Postgres database logs, containing statements executed by connected applications.
- `auth_logs`: GoTrue server logs, containing authentication/authorization activity.
- `storage_logs`: Storage server logs, containing object upload and retrieval information.
- `realtime_logs`: Realtime server logs, containing client connection information.
- `function_edge_logs`: Edge network logs for only edge functions, containing network requests and response metadata for each execution.
- `function_logs`: Function internal logs, containing any `console` logging from within the edge function.

## Querying with the Logs Explorer

> The underlying SQL engine used for the Logs Explorer is BigQuery, hence you will be able to use any of the [available SQL functions and operators]((https://cloud.google.com/bigquery/docs/reference/standard-sql/functions-and-operators)) supported by BigQuery.

Reviewer: ```suggestion
The underlying SQL engine for the Logs Explorer is BigQuery. You can use any of the [available SQL functions and operators]((https://cloud.google.com/bigquery/docs/reference/standard-sql/functions-and-operators)) supported by BigQuery.
```

### Timestamp Display and Behavior

> Each log entry is stored with a `timestamp`. In order to utilize the `timestamp` field in a query, you can use the appropriate [timestamp functions](https://cloud.google.com/bigquery/docs/reference/standard-sql/timestamp_functions#timestamp).

Reviewer: ```suggestion
Each log entry is stored with a `timestamp`. To utilize the `timestamp` field in a query, you can use [timestamp functions](https://cloud.google.com/bigquery/docs/reference/standard-sql/timestamp_functions#timestamp).
```

> Raw top-level timestamp values are rendered as unix microsecond timestamps. SQL queries, however, should always use the `TIMESTAMP` data type. If you are using a unix timestamp value in a query, cast the value to a `TIMESTAMP` data type.

Reviewer: might need an example

Author: added

To render the timestamps in a human-readable format, we recommend using the `DATETIME()` function to convert the unix timestamp display into an ISO-8601 timestamp.

```sql
SELECT DATETIME(timestamp) FROM ....
```

### Unnesting Arrays

> To query nested data within log event, you will need to unnest the field and "join" the unnested data. Clicking on the log row shows that log metadata is stored as an array of objects.

Reviewer: ```suggestion
To query nested data within log events, unnest the field and "join" the unnested data. Clicking on the log row shows that log metadata is stored as an array of objects.
```

Reviewer: this definitely needs an example or a screenshot

> In order to query any value that is an array, we would need to `UNNEST()` that field and add it to the query as a join, thereby allowing us to reference the nested fields within the array.

Reviewer: ```suggestion
To query arrays, use `UNNEST()` on the array field and join it to the query. This allows you to reference the nested fields within the array.
```

> For example, a log event's metadata is stored as a repeated `RECORD` within BigQuery, hence we would need to do the following in order to query across the nested fields:

Reviewer: ```suggestion
For example, a log event's metadata is stored as a repeated `RECORD` within BigQuery. You can query this data with `UNNEST()`:
```

```sql
SELECT timestamp, r.method, h.cf_ipcountry
FROM edge_logs t
CROSS JOIN UNNEST(t.metadata) as m
CROSS JOIN UNNEST(m.request) as r
CROSS JOIN UNNEST(r.headers) as h
```

> Here, the `metadata` key when rendered in the Logs Explorer shows an array. In order to drill down further, we perform the `CROSS JOIN UNNEST()` to select the nested key `request`. However, in order to query the headers, we have to go one step further and perform the join and unnest again in order to have access to the headers, under the alias `h`.

Reviewer: ```suggestion
Here, the `metadata` is an array. We perform the `CROSS JOIN UNNEST()` to select the nested key `request`. To query the headers, we have to go one step further and perform the join and unnest again in order to have access to the headers, under the alias `h`.
```

> Here, the `metadata` key when rendered in the Logs Explorer shows an array. In order to drill down further, we perform the `CROSS JOIN UNNEST()` to select the nested key `request`. However, in order to query the headers, we have to go one step further and perform the join and unnest again in order to have access to the headers, under the alias `h`.

Reviewer 2: ```suggestion
The `metadata` key is displayed as an array in the Logs Explorer. In order to drill down further, perform the `CROSS JOIN UNNEST()` to select the nested key `request`. In order to query the headers, perform the join and unnest again under the alias `h`.
```

### LIMIT and Result Row Limitations

> The Logs Explorer has a maximum of 1000 rows per run. To optimize your queries, you may wish to add in a `LIMIT` to your SQL query, to reduce the number of rows returned even further.

Reviewer: ```suggestion
The Logs Explorer has a maximum of 1000 rows per run. Use `LIMIT` optimize your queries by reducing the number of rows returned even further
```

> The Logs Explorer has a maximum of 1000 rows per run. To optimize your queries, you may wish to add in a `LIMIT` to your SQL query, to reduce the number of rows returned even further.

Reviewer 2: ```suggestion
The Logs Explorer has a maximum of 1000 rows per run. To optimize your queries, you can add a `LIMIT` to your SQL query to reduce the number of rows returned.
```

### Best Practices

> 1. Have a WHERE statement filtering over **timestamp**

Reviewer: ```suggestion
1. Always filter by `timestamp`
```

> 1. Have a WHERE statement filtering over **timestamp**

Reviewer 2: ```suggestion
1. Include a WHERE statement filtering over **timestamp**
```

>   Querying across your entire log retention range might seem appealing, especially when you just need to get your query results fast. However, for **Enterprise** customers that have a large retention range, you would run the risk of your queries timing out due to the length cruch time required to scan and analyze across your large dataset. We recommend optimizing queries based on timestamp filtering, to ensure

Reviewer: ```suggestion
  Querying your entire log history might sound appealing. However, for **Enterprise** customers with a large log history, you run the risk of timeouts due to the time required to analyze such a large dataset. We recommend optimizing queries based on timestamp filtering, to ensure
```

to ensure what?

>   Querying across your entire log retention range might seem appealing, especially when you just need to get your query results fast. However, for **Enterprise** customers that have a large retention range, you would run the risk of your queries timing out due to the length cruch time required to scan and analyze across your large dataset. We recommend optimizing queries based on timestamp filtering, to ensure

Reviewer 2: Last sentence ends abruptly...
```suggestion
Querying across your entire log retention range might seem appealing, especially when you just need to get your query results fast. However, for Enterprise customers that have a large retention range, you run the risk of your queries timing out due to the length of time required to scan and analyze across your large dataset.
```

2. Avoid selecting large nested objects. Select individual values instead.

>   When querying large object, the columnar storage engine (in this case BigQuery) will select each column associated with each nested key, resulting in a large number of columns being selected. This inadvertently impacts the query speed and may result in timeouts or memory errors, especially for projects with large amount of logging activity. 

Reviewer: ```suggestion
  When querying large objects, the columnar storage engine selects each column associated with each nested key, resulting in a large number of columns being selected. This impacts the query speed and may result in timeouts or memory errors, especially for projects with a lot of logs. 
```

>   When querying large object, the columnar storage engine (in this case BigQuery) will select each column associated with each nested key, resulting in a large number of columns being selected. This inadvertently impacts the query speed and may result in timeouts or memory errors, especially for projects with large amount of logging activity. 

Reviewer 2: ```suggestion
  When querying large objects, the columnar storage engine (in this case BigQuery) selects each column associated with each nested key, resulting in a large number of columns being selected. This inadvertently impacts the query speed and may result in timeouts or memory errors, especially for projects with large amounts of logging activity. 
```

>   Instead, try to select only the values required, and where this is not possible, drill down to the minimally required object keys.

Reviewer: ```suggestion
  Instead, select only the values required.
```

  ```sql
  -- ❌ edge_logs metadata contains many nested keys
  SELECT timestamp, m as metadata
  FROM edge_logs t

>   CROSS JOIN UNNEST(t.metadata) as m;

Reviewer 2: Thoughts about commenting out the actual statement too so it's more apparent as something to not do?

Author: Hmm, the syntax highlighting would make it more readable though 🤔 
I've shifted the comment on the nested keys to be inline, and added a more obvious "Avoid doing this" comment

  
  -- ✅ select only the required values
  SELECT timestamp, r.method
  FROM edge_logs t
  CROSS JOIN UNNEST(t.metadata) as m
  CROSS JOIN UNNEST(m.request) as r
  ```

### Examples and Templates

> For example, you may enter the following into the SQL editor to query for each user's IP address:

Reviewer: "For example"...

This doesn't seem to be the right place for this sentence

Author: shifted another para before this sentence

```sql

> SELECT DATETIME(timestamp), h.x_real_ip

Reviewer: can we please use lower-case everywhere for SQL? It's more approachable for newcomers to SQL

FROM edge_logs
  LEFT JOIN UNNEST(metadata) as m ON TRUE
  LEFT JOIN UNNEST(m.request) AS r ON TRUE
  LEFT JOIN UNNEST(r.headers) AS h ON TRUE
WHERE h.x_real_ip IS NOT NULL AND r.method = "GET"
```

> Within the Logs Explorer, we have also included some templates to help you get started on each. Templates are available under the **Templates** tab, or under the **Templates** Dropdown in the **Query** tab.

Reviewer: ```suggestion
Within the Logs Explorer, we have included templates. Templates are available under the **Templates** tab, or under the **Templates** Dropdown in the **Query** tab.
```

> For example, you may enter the following into the SQL editor to query for each user's IP address:
> 
> ```sql
> SELECT DATETIME(timestamp), h.x_real_ip
> FROM edge_logs
>   LEFT JOIN UNNEST(metadata) as m ON TRUE
>   LEFT JOIN UNNEST(m.request) AS r ON TRUE
>   LEFT JOIN UNNEST(r.headers) AS h ON TRUE
> WHERE h.x_real_ip IS NOT NULL AND r.method = "GET"
> ```
> 
> Within the Logs Explorer, we have also included some templates to help you get started on each. Templates are available under the **Templates** tab, or under the **Templates** Dropdown in the **Query** tab.

Reviewer 2: ````suggestion
The Logs Explorer includes **Templates** (available in the Templates tab or the dropdown in the Query tab) to help you get started.

For example, you can enter the following query in the SQL Editor to retrieve each user's IP address:

```sql
SELECT DATETIME(timestamp), h.x_real_ip
FROM edge_logs
  LEFT JOIN UNNEST(metadata) as m ON TRUE
  LEFT JOIN UNNEST(m.request) AS r ON TRUE
  LEFT JOIN UNNEST(r.headers) AS h ON TRUE
WHERE h.x_real_ip IS NOT NULL AND r.method = "GET"
```
````

## Limitations

> - Logging UI is currently not available for self-hosted and local development. This is currently on the roadmap and you can follow the progress in the [Logflare repository](https://github.com/Logflare/logflare).

Reviewer: ```suggestion
- The Logging interface is currently not available for self-hosted and local development. This is on the roadmap and you can follow the progress in the [Logflare repository](https://github.com/Logflare/logflare).
```

> - Unoptimized queries for projects with high logging activity may lead to memory errors or timeouts. Please refer to the [best practices](#best-practices) when crafting your queries.

Reviewer 2: Hmm, I feel like this section is a bit redundant. Suggest moving the first bullet point to the callout at the top of the doc. The second bullet point is unnecessary—assuming the user has made it to this point in the doc, they've likely seen the best practices section.

Author: 👍

description: Getting started with Supabase Platform Log Browser
---

import Tabs from '@theme/Tabs'
import TabItem from '@theme/TabItem'

The Supabase Platform includes a Logs Explorer that allows log tracing and debugging. Log retention is based on your [project's pricing plan](https://supabase.com/pricing).

:::note

> These features are currently not available for self-hosted and local development.<br/>
> This is currently on the roadmap and you can follow the progress in the [Logflare repository](https://github.com/Logflare/logflare).

Reviewer 2: ```suggestion
These features are not currently available for self-hosting and local development.<br/>
This is on the roadmap and you can follow the progress in the [Logflare repository](https://github.com/Logflare/logflare).
```

:::

## Product Logs

> Supabase provides a logging interface specific to each product. Simple regex searching is available through the search bar to search the log event messsage for specific keywords and patterns. You may export and download the log events matching your query as a spreadsheet.

Reviewer 2: ```suggestion
Supabase provides a logging interface specific to each product. You can use simple regular expressions for keywords and patterns to search log event messages. You can also export and download the log events matching your query as a spreadsheet.
```

<!--
To update the screenshots, ensure that at least one log line is selected to display the metadata. Can use meme.town as an example.
-->

<Tabs>
  <TabItem value="api" label="API" default>

[API logs](https://app.supabase.com/project/_/database/api-logs) show all network requests and response for the REST and GraphQL [APIs](../../guides/api).

![API Logs](/img/guides/platform/logs/logs-api.png)

  </TabItem>
  <TabItem value="postgres" label="Postgres">

[Postgres logs](https://app.supabase.com/project/_/database/postgres-logs) show all queries and activity for your [database](../../guides/database).

![Postgres Logs](/img/guides/platform/logs/logs-database.png)

  </TabItem>
  <TabItem value="auth" label="Auth">

> [Auth logs](https://app.supabase.com/project/_/auth/logs) show all server logs for your [Auth uage](../../guides/auth).

Reviewer 2: ```suggestion
[Auth logs](https://app.supabase.com/project/_/auth/logs) show all server logs for your [Auth usage](../../guides/auth).
```

- `realtime_logs`: Realtime server logs, containing client connection information.
- `storage_logs`: Storage server logs, containing object upload and retrieval information.

## Querying with the Logs Explorer

The Logs Explorer uses BigQuery and supports all [available SQL functions and operators](https://cloud.google.com/bigquery/docs/reference/standard-sql/functions-and-operators).

### Timestamp Display and Behavior

Each log entry is stored with a `timestamp` as a `TIMESTAMP` data type. Use the appropriate [timestamp function](https://cloud.google.com/bigquery/docs/reference/standard-sql/timestamp_functions#timestamp) to utilize the `timestamp` field in a query.

Raw top-level timestamp values are rendered as unix microsecond. To render the timestamps in a human-readable format, use the `DATETIME()` function to convert the unix timestamp display into an ISO-8601 timestamp.

```sql
-- timestamp column without datetime()
select timestamp from ....
--  1664270180000

-- timestamp column with datetime()
select datetime(timestamp) from ....
-- 2022-09-27T09:17:10.439Z
```

:::info
The rendering behaviour as unix microsecond will be removed in the future.

> :::

Reviewer 2: Suggest removing this unless it's absolutely necessary to have. Docs should avoid mentioning anything in the future.

### Unnesting Arrays

Each log event stores metadata an array of objects with multiple levels, and can be seen by selecting single log events in the Logs Explorer. To query arrays, use `unnest()` on each array field and add it to the query as a join. This allows you to reference the nested objects with an alias and select their individual fields.

For example, to query the edge logs without any joins:

```sql
select timestamp, metadata from edge_logs t
```

> The resulting `metadata` key is rendered as an array of objects when rendered in the Logs Explorer. This can be better visualized in this diagram, where each square in a column represents a nested object:

Reviewer 2: Per the Slack thread:
```suggestion
The resulting `metadata` key is rendered as an array of objects in the Logs Explorer. In the following diagram, each box within the metadata column represents a nested object.
```

> ![Without Unnesting](/img/guides/platform/logs/unnesting-none.png)

Reviewer 2: These diagrams aren't visible in dark mode:
<img width="902" alt="Screen Shot 2022-09-27 at 12 18 30 PM" src="https://user-images.githubusercontent.com/7026076/192593671-d46ff1d8-c67b-4f10-848a-4921b77cc566.png">

Reviewer 3: I'm assuming these diagrams are placeholder.
if we were intending to ship these, we can do better 👀🚀

Are these diagrams finalized on what they're communicating?

Reviewer 3: we can also maybe used actual dashboard screenshots so there's context that it's happening in the dashboard

Author: I've switched these to use excalidraw instead.

These are conceptual diagrams for user understanding, not related to actual UI.

I've added image switching based on the theme

<img width="989" alt="Screenshot 2022-09-28 at 7 17 44 PM" src="https://user-images.githubusercontent.com/22714384/192776741-6d5148b2-f77d-4fd6-8679-cd55e1386d94.png">
<img width="1026" alt="Screenshot 2022-09-28 at 7 17 40 PM" src="https://user-images.githubusercontent.com/22714384/192776751-f134157a-343c-44fa-b36b-5205f8ea68c4.png">

<img width="1084" alt="Screenshot 2022-09-28 at 7 17 37 PM" src="https://user-images.githubusercontent.com/22714384/192776766-a82ae07f-ab06-4894-a2f1-e43af0097aff.png">
<img width="1165" alt="Screenshot 2022-09-28 at 7 17 33 PM" src="https://user-images.githubusercontent.com/22714384/192776780-907f7369-fcdb-43d9-aaee-d698c7d46511.png">

> We can perform a `cross join unnest()` to work with keys nested below the `metadata` key.

Reviewer 2: Style nit: Avoid using "we" and "us" as that implies a relationship between Supabase and the user, but there isn't really one (i.e., we not building something or going through a flow together with the user). Default to "you" if you're unable to start a sentence with an action/verb like the suggestion below.
```suggestion
Perform a `cross join unnest()` to work with the keys nested in the `metadata` key.
```

> To query for a deeply nested value, we will have to add a join for each array level:

Reviewer 2: Would this still be correct?
```suggestion
To query for a nested value, add a join for each level of the array:
```

Author: will keep the array wording, should be "each array level", as there are many nested arrays (i.e. arrrays within arrays within arrays)

```sql
select timestamp, request.method, header.cf_ipcountry
from edge_logs t
cross join unnest(t.metadata) as metadata
cross join unnest(metadata.request) as request
cross join unnest(request.headers) as header
```

> Doing so surfaces the following columns available for selection:

Reviewer 2: ```suggestion
This surfaces the following columns available for selection:
```

![With Two Level Unnesting](/img/guides/platform/logs/unnesting-2.png)

> This then allows us to select the `method` and `cf_ipcountry` columns. In JS dot notation, the full paths for each selected column are:

Reviewer 2: ```suggestion
This allows you to select the `method` and `cf_ipcountry` columns. In JS dot notation, the full paths for each selected column are:
```

    dark: useBaseUrl("/img/guides/platform/logs/unnesting-none-dark.png"),
  }}
  style={{maxHeight: 200}}
/>

Perform a `cross join unnest()` to work with the keys nested in the `metadata` key.

To query for a nested value, add a join for each array level:

```sql
select timestamp, request.method, header.cf_ipcountry
from edge_logs t
cross join unnest(t.metadata) as metadata
cross join unnest(metadata.request) as request
cross join unnest(request.headers) as header
```

This surfaces the following columns available for selection:

<ThemedImage
  alt="With Two Level Unnesting"
  sources={{
    light: useBaseUrl("/img/guides/platform/logs/unnesting-2.png"),
    dark: useBaseUrl("/img/guides/platform/logs/unnesting-2-dark.png"),
  }}
  style={{maxHeight: 200}}
/>

This allows you to select the `method` and `cf_ipcountry` columns. In JS dot notation, the full paths for each selected column are:

- `metadata[].request[].method`
- `metadata[].request[].headers[].cf_ipcountry`

### LIMIT and Result Row Limitations

The Logs Explorer has a maximum of 1000 rows per run. Use `LIMIT` to optimize your queries by reducing the number of rows returned further.

### Best Practices

> 1. Include a filter over **timestamp**
> 
> Querying your entire log history might seem appealing. For **Enterprise** customers that have a large retention range, you run the risk of timeouts due additional time required to scan the larger dataset.
> 
> 2. Avoid selecting large nested objects. Select individual values instead.
> 
> When querying large objects, the columnar storage engine selects each column associated with each nested key, resulting in a large number of columns being selected. This inadvertently impacts the query speed and may result in timeouts or memory errors, especially for projects with a lot of logs.
> 
> Instead, select only the values required.
> 
> ```sql
> -- ❌ Avoid doing this
> select
>   datetime(timestamp),
>   m as metadata  -- <- metadata contains many nested keys
> from edge_logs t
> cross join unnest(t.metadata) as m;
> 
> -- ✅ Do this
> select
> datetime(timestamp),
> r.method   -- <- select only the required values
> from edge_logs t
> cross join unnest(t.metadata) as m
> cross join unnest(m.request) as r
> ```

Reviewer 2: Indent the content within each best practice. Can you also double check the indentation for the 2nd SQL example?
````suggestion
1. Include a filter over **timestamp**.

  Querying your entire log history might seem appealing. For **Enterprise** customers that have a large retention range, you run the risk of timeouts due additional time required to scan the larger dataset.

2. Avoid selecting large nested objects. Select individual values instead.

  When querying large objects, the columnar storage engine selects each column associated with each nested key, resulting in a large number of columns being selected. This inadvertently impacts the query speed and may result in timeouts or memory errors, especially for projects with a lot of logs. Instead, select only the values required.

  ```sql
  -- ❌ Avoid doing this
  select
    datetime(timestamp),
    m as metadata  -- <- metadata contains many nested keys
  from edge_logs t
  cross join unnest(t.metadata) as m;

  -- ✅ Do this
  select
  datetime(timestamp),
  r.method   -- <- select only the required values
  from edge_logs t
  cross join unnest(t.metadata) as m
  cross join unnest(m.request) as r
  ```
````
